warehouse/redshift.go

```diff
@@ -110,14 +110,19 @@ func (rs *Redshift) UploadFile(name string) (string, error) {
 	defer cancelFn()
 
 	_, objName := filepath.Split(name)
-	s3path := fmt.Sprintf("s3://%s/%s", rs.conf.S3.Bucket, objName)
+
+	bucketParts := strings.Split(rs.conf.S3.Bucket, "/")
+	bucketName := bucketParts[0]
+	keyPath := strings.Trim(strings.Join(bucketParts[1:], "/"), "/")
+	key := fmt.Sprintf("%s/%s", keyPath, objName)
```

Comment:
What this block is doing is subtle enough that it would be great to have a test for it, both to serve as documentation and to make sure we don't inadvertently break this in the future. You could refactor this logic into a function like so:

```
func getBucketAndKey(bucketPath, objName string) (string, string){
	bucketParts := strings.Split(bucketPath, "/")
	keyPath := strings.Trim(strings.Join(bucketParts[1:], "/"), "/")
	key := fmt.Sprintf("%s/%s", keyPath, objName)
	return bucketParts[0], key
}
```

Then in `redshift_test.go`, add a test function similar to this:

```
func TestGetBucketAndKey(t *testing.T) {
	testCases := []struct {
		s3Config string
		fileName string
		expBucket string
		expKey string
	} {
		{
			s3Config: "plainbucket",
			fileName: "data.csv",
			expBucket: "plainbucket",
			expKey: "data.csv",
		},
		{
			s3Config: "withslash/",
			fileName: "data.csv",
			expBucket: "withslash",
			expKey: "data.csv",
		},
		// TODO: test cases around subdirectories
	}
	
	for i, tc := range testCases {
		// TODO: call getBucketAndKey with tc.sConfig and tc.fileName
		// TODO: compare result with expected values, t.Errorf if they don't match
	}
}
```

Reply:
Done!